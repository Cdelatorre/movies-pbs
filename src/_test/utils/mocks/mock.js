export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
