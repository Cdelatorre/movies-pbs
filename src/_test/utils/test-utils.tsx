import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
