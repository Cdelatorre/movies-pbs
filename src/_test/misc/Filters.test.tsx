import React from "react";
import { screen } from "@testing-library/react";
import { renderWithProviders } from "../utils/test-utils";
import Filters from "../../components/misc/filters/index";
import { sampleFiltersData } from "../utils/mocks/mock";

describe("Renders Filters component correctly", () => {
  test("Renders Filters default component correctly", () => {
    renderWithProviders(
      <Filters filters={sampleFiltersData} activeFilters={[]} />
    );

    expect(screen.getByTestId("filters-component")).toBeInTheDocument();
    sampleFiltersData.forEach((genre) => {
      expect(screen.getByTestId(`input-checkbox-${genre}`)).toBeInTheDocument();
    });
    expect(screen.getByTestId("test-search-btn")).toBeInTheDocument();
    expect(screen.getByTestId("search-bar")).toBeInTheDocument();
  });

  test("Inputs are checked if filters includes some activeFilters", () => {
    const activeFilters = ["Horror", "Comedy"];
    renderWithProviders(
      <Filters filters={sampleFiltersData} activeFilters={activeFilters} />
    );

    activeFilters.forEach((genre) => {
      const input = screen.getByTestId(`input-checkbox-${genre}`);
      expect(input).toHaveAttribute("checked");
    });

    sampleFiltersData
      .filter((f) => !activeFilters.includes(f))
      .forEach((genre) => {
        const input = screen.getByTestId(`input-checkbox-${genre}`);
        expect(input).not.toHaveAttribute("checked");
      });
  });
});
