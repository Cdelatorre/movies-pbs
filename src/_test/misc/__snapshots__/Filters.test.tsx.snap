// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Filters component correctly Renders Filters default component correctly 1`] = `
<div
  class="filters-container d-flex flex-column align-items-start"
  data-testid="filters-component"
>
  <h2
    class="text-light text-start mb-4"
  >
    Find your movie!
  </h2>
  <div
    class="sub-decoration"
  />
  <div
    class="d-flex flex-wrap"
  >
    <div
      class="d-flex text-light flex-row flex-wrap w-100"
    >
      <div
        class="input form-check"
      >
        <label
          class="d-inline ps-2 pe-3 text-start text-light"
          for="Horror"
        >
          Horror
        </label>
        <input
          class="form-control form-checkbox bg-dark text-light"
          data-testid="input-checkbox-Horror"
          id="Horror"
          name="Horror"
          readonly=""
          type="checkbox"
          value="Horror"
        />
      </div>
      <div
        class="input form-check"
      >
        <label
          class="d-inline ps-2 pe-3 text-start text-light"
          for="Comedy"
        >
          Comedy
        </label>
        <input
          class="form-control form-checkbox bg-dark text-light"
          data-testid="input-checkbox-Comedy"
          id="Comedy"
          name="Comedy"
          readonly=""
          type="checkbox"
          value="Comedy"
        />
      </div>
      <div
        class="input form-check"
      >
        <label
          class="d-inline ps-2 pe-3 text-start text-light"
          for="Romance"
        >
          Romance
        </label>
        <input
          class="form-control form-checkbox bg-dark text-light"
          data-testid="input-checkbox-Romance"
          id="Romance"
          name="Romance"
          readonly=""
          type="checkbox"
          value="Romance"
        />
      </div>
      <button
        class="btn btn-primary"
        data-testid="test-reset-filters-btn"
        id="reset-filters-btn"
      >
        <i
          class="fas fa-sync-alt"
        />
      </button>
    </div>
  </div>
  <div
    class="search-bar mb-5 mt-4 w-100 d-flex"
    data-testid="search-bar"
  >
    <div
      class="input "
    >
      <input
        class="form-control form-text bg-dark text-light w-100"
        data-testid="input-text-search"
        name="search"
        placeholder="Search movie"
        type="text"
        value=""
      />
    </div>
    <button
      class="btn btn-success ms-3"
      data-testid="test-search-btn"
      id="search-btn"
    >
      Search
    </button>
  </div>
</div>
`;
